src/Data/Fin/Relation/Unary/Top.agda

```diff
@@ -0,0 +1,257 @@
+
+view : ∀ {n} i → View {n} i
+view {n = zero}  zero = top
+view {n = suc _} zero = inj _
+view {n = suc n} (suc i) with view {n} i
+... | top   = top
+... | inj j = inj (suc j)
+
+-- Interpretation of the view constructors
+
+⟦_⟧ : ∀ {i} → View {n} i → Fin (suc n)
+⟦ top ⟧   = fromℕ _
+⟦ inj j ⟧ = inject₁ j
+
+-- Completeness of the view
+
+view-complete : ∀ {i} (v : View {n} i) → ⟦ v ⟧ ≡ i
+view-complete top     = refl
+view-complete (inj j) = refl
+
+-- 'Computational' behaviour of the covering function
+view-top : ∀ n → view {n} (fromℕ n) ≡ top
+view-top zero    = refl
+view-top (suc n) rewrite view-top n = refl
+
+view-inj : ∀ j → view {n} (inject₁ j) ≡ inj j
+view-inj zero       = refl
+view-inj (suc j) rewrite view-inj j = refl
+
+------------------------------------------------------------------------
+-- Experimental
+--
+-- Because we work --without-K, Agda's unifier will complain about
+-- attempts to match `refl` against hypotheses of the form
+--    `view {n] i ≡ v`
+-- or gets stuck trying to solve unification problems of the form
+--    (inferred index ≟ expected index)
+-- even when these problems are *provably* solvable.
+--
+-- So the two predicates on values of the view defined below
+-- are extensionally equivalent to the assertions
+-- `view {n] i ≡ top` and `view {n] i ≡ inj j`
+--
+-- But such assertions can only ever have a unique (irrelevant) proof
+-- so we introduce instances to witness them, themselves given by
+-- the functions `view-top` and `view-inj` defined above
+
+module Instances {n} where
+
+  data IsTop : ∀ {i} → View {n} i → Set where
+
+    top : IsTop top
+
+  instance
+
+    top⁺ : IsTop {i = fromℕ n} (view (fromℕ n))
+    top⁺ rewrite view-top n = top
+
+  data IsInj : ∀ {i} → View {n} i → Set where
+
+    inj : ∀ j → IsInj (inj j)
+
+  instance
+
+    inj⁺ : ∀ {j} → IsInj (view (inject₁ j))
+    inj⁺ {j} rewrite view-inj j = inj j
+
+    inject₁≡⁺ : ∀ {i} {j} {eq : inject₁ i ≡ j} → IsInj (view j)
+    inject₁≡⁺ {eq = refl} = inj⁺
+
+    inject₁≢n⁺ : ∀ {i} {n≢i : n ≢ toℕ (inject₁ i)} → IsInj (view {n} i)
+    inject₁≢n⁺ {i} {n≢i} with view i
+    ... | top   = ⊥-elim (n≢i (begin
+      n                         ≡˘⟨ toℕ-fromℕ n ⟩
+      toℕ (fromℕ n)           ≡˘⟨ toℕ-inject₁ (fromℕ n) ⟩
+      toℕ (inject₁ (fromℕ n)) ∎)) where open ≡-Reasoning
+    ... | inj j = inj j
+
+open Instances
+
+------------------------------------------------------------------------
+-- As a corollary, we obtain two useful properties, which are
+-- witnessed by, but can also serve as proxy replacements for,
+-- the corresponding properties in `Data.Fin.Properties`
+
+module _ {n} where
+
+  view-top-toℕ : ∀ i → .{{IsTop (view i)}} → toℕ i ≡ n
+  view-top-toℕ i with top ← view i = toℕ-fromℕ n
+
+  view-inj-toℕ< : ∀ i → .{{IsInj (view i)}} → toℕ i < n
+  view-inj-toℕ< i with inj j ← view i = inject₁ℕ< j
+
+------------------------------------------------------------------------
+-- Examples
+
+private module Examples {n} where
```

Comment:
I **think** that people have put such examples under the `README.` hierarchy. This is valuable stuff and shouldn't be lost, but I dislike polluting the main source file with this. [To be clear: I would rather have it remain here than it be deleted.]

Reply:
OK, I haven't really experimented with `README` files before... and this was really the 'discussion' part of the PR: using views, we can recreate all of what has gone before, and with more or less (mostly less!  I guess YMMV) work. 

Elevating it to `README` feels like trying to have it both ways: retaining the 'old' way, but showing how to use the new? But happy to be nudged either way... and I think some more discussion with others would be helpful.

Reply:
Will work on this later this weekend.

Reply:
I'm personally of the opinion that `Instances` should not be `public`, but it's @MatthewDaggitt 's call as to what the convention is in the overall library.

In a `README` file, since the intent is pedagogical instead of being "library standard", I think there should be a lot more leeway. So there, I'd go with whatever you think illustrates the idea most clearly.

Reply:
Good, thanks. I think I agree with making `Instances` *not* public.

As for `opposite`, I think the name duplication does make the point forcefully. 
For `inject₁⁻¹`, even more so, as we no longer have to think what the relationship is between `lower₁` and `inject₁`.